File: assets/static/blog/dsa/data_structures/linked_lists/implementation.go
package main
import "fmt"

type Node struct {
  value interface{}
  next *Node
}

type LinkedList struct {
  head *Node
  size int
}

func (l *LinkedList) Prepend(val interface{}) {
  head := l.head
  l.head = &Node{value: val, next: head}
  l.size += 1
}

func (l *LinkedList) Append(val interface{}) {
  if l.size == 0 {
    l.head = &Node{value: val}
    l.size += 1
    return
  }

  node := l.head
  for i := 0; i < l.size-1; i++ {
    node = node.next
  }
  node.next = &Node{value: val}
  l.size += 1
}

func (l *LinkedList) PopAt(ind int) interface{} {

  if ind == 0 {
    val := l.head.value
    l.head = l.head.next
    l.size -= 1
    return val
  }

  prev := &Node{}
  node := l.head
  for i := 0; i < ind; i++ {
    prev = node
    node = node.next
  }

  val := node.value
  prev.next = node.next
  l.size -= 1
  return val
}

func (l *LinkedList) Reverse() {
  prev := &Node{}
  next := &Node{}
  curr := l.head
  for i := 0; i < l.size; i++ {
    next = curr.next
    curr.next = prev
    prev = curr
    curr = next
  }
  l.head = prev
}

func (l *LinkedList) DisplayValues() {
  node := l.head
  for i := 0; i < l.size; i++ {
    fmt.Print(node.value, ", ")
    node = node.next
  }
  fmt.Println()
}



func main() {
  l := LinkedList{}
  l.Append("there")
  l.Prepend("hi")
  l.Append("small")
  l.Append("gerbil")
  l.PopAt(2)
  l.DisplayValues()

  l.Reverse()
  l.DisplayValues()

}

File: assets/static/blog/dsa/data_structures/linked_lists/implementation.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None

    def __repr__(self):
        return str(self.data)


class LinkedList:
    def __init__(self, head=None):
        self.head = None
        self.size = 1 if head else 0

    def __len__(self):
        return self.size

    def __iter__(self):
        self._node = Node()
        self._node.next = self.head
        return self

    def __next__(self):
        if not self._node.next:
            raise StopIteration()
        self._node = self._node.next
        return self._node

    def __getitem__(self, ind):
        if not self.head:
            raise IndexError("List is empty")
        if ind > self.size-1:
            raise IndexError("Index out of bounds")
        if ind < 0:
            ind = self.size + ind
            if ind < 0:
                raise IndexError("Index out of bounds")

        for i, node in enumerate(self):
            if i == ind:
                return node

    def __repr__(self):
        if not self.head:
            return "[]"
        r = "["
        for i, node in enumerate(self):
            r += f"{node.data}, "
        return r[:-2] + ']'

    @property
    def tail(self):
        return self[-1]

    def append(self, item):
        if not self.head:
            self.head = Node(item)
        else:
            self.tail.next = Node(item)
        self.size += 1

    def prepend(self, item):
        if not self.head:
            self.head = Node(item)
        else:
            new_head = Node(item)
            new_head.next = self.head
            self.head = new_head
        self.size += 1

    def pop_first(self):
        if not self.head:
            return None
        node = self.head
        self.head = self.head.next
        self.size -= 1
        return node

    def pop_last(self):
        if not self.head:
            return None
        if self.head.next is None:
            return self.pop_head()

        new_tail = self[-2]
        tail = new_tail.next
        new_tail.next = None
        self.size -= 1
        return tail

    def pop_at(self, ind):
        if ind == 0:
            return self.pop_first()
        if ind == self.size - 1:
            return self.pop_last()

        prev_node = self[ind-1]
        node = prev_node.next
        prev_node.next = node.next
        return node

    def insert(self, ind, item):
        if ind == 0:
            self.prepend(item)
            return

        node = self[ind-1]
        nxt = node.next
        node.next = Node(item)
        node.next.next = nxt
        self.size += 1

    def reverse(self):
        if not self.head:
            return None

        prev = None
        current = self.head
        while current is not None:
            next = current.next
            current.next = prev
            prev = current
            current = next
        self.head = prev


def test():
    l = LinkedList()
    l.append(1)
    l.append(2)
    l.append(3)
    l.append(4)
    l.prepend(1)
    l.prepend(2)
    l.prepend(3)
    l.prepend(4)
    l.pop_first()
    l.pop_last()
    l.pop_at(0)
    l.pop_at(0)
    l.pop_at(0)

    print(l)
    l.reverse()
    l.insert(2, "hi")
    print(l)
    return l

l = test()
l
l[3]
